### PR TITLE
Implement the AWS Query serialization codegen

### PR DESCRIPTION
**Pull request by jdisanti**
Implements serialization for #222.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/AwsQuerySerializerGenerator.kt [hunk 1]
```diff
@@ -1,56 +1,304 @@
...
+        val structureSerializer = RuntimeType.forInlineFun(fnName, "query_ser") { writer ->
+            val writerMut = when (topLevel) {
+                true -> ""
+                else -> "mut "
+            }
+            val writerType = when (topLevel) {
+                true -> "&mut #{QueryWriter}"
+                false -> "#{QueryValueWriter}"
+            }
+            Attribute.AllowUnusedMut.render(writer)
```
Comment (rcoh) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/AwsQuerySerializerGenerator.kt]: I've done worse, but in the past I've dealt with this by having `serializeStructureInner`. When it's used for operations, we inline `serializeStructureInner`.

Serialize structure sets up a function signature then calls `serializeStructureInner`

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/AwsQuerySerializerGenerator.kt [hunk 2]
```diff
@@ -1,56 +1,304 @@
...
+        when (target) {
+            is StringShape -> when (target.hasTrait<EnumTrait>()) {
+                true -> rust("$writer.string(${value.name}.as_str());")
+                false -> rust("$writer.string(${value.name});")
+            }
+            is BooleanShape -> rust("$writer.boolean(${value.asValue()});")
+            is NumberShape -> {
+                val numberType = when (symbolProvider.toSymbol(target).rustType()) {
+                    is RustType.Float -> "Float"
+                    is RustType.Integer -> "NegInt"
```
Comment (rcoh) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/AwsQuerySerializerGenerator.kt:187]: is this right?

Comment (rcoh) [reply codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/AwsQuerySerializerGenerator.kt:187]: ohhh it generates i64? Can you comment that?

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/AwsQuerySerializerGenerator.kt [hunk 3]
```diff
@@ -1,56 +1,304 @@
...
+        safeName("map").also { mapName ->
+            val keyName = safeName("key")
+            val valueName = safeName("value")
+            rust("let mut $mapName = ${context.writerExpression}.start_map($flat, $entryKeyName, $entryValueName);")
+            rustBlock("for ($keyName, $valueName) in ${context.valueExpression.asRef()}") {
+                val keyTarget = model.expectShape(context.shape.key.target)
+                val keyExpression = when (keyTarget.hasTrait<EnumTrait>()) {
+                    true -> "$keyName.as_str()"
+                    else -> keyName
+                }
```
Comment (rcoh) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/AwsQuerySerializerGenerator.kt:263]: at some point I had a helper that literally just did this for enums, but I guess I deleted it

File: rust-runtime/protocol-test-helpers/src/lib.rs [hunk 4]
```diff
@@ -310,11 +321,58 @@ fn try_json_eq(actual: &str, expected: &str) -> Result<(), ProtocolTestFailure>
     }
 }
 
+fn rewrite_url_encoded_map_keys(input: &str) -> (String, String) {
```
Comment (rcoh) [inline rust-runtime/protocol-test-helpers/src/lib.rs:324]: since this file is already kinda mega, can you put this in its own module like XML?

**Review by rcoh — APPROVED** [review]
LGTM! One thing to reduce future pain: Can you run your unit test with Kotlin coverage on to see if there are missing branches that we don't generate code for?

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/AwsQuerySerializerGenerator.kt (hunk 1 above)
Comment (jdisanti) [reply codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/AwsQuerySerializerGenerator.kt]: This is fixed.

File: rust-runtime/protocol-test-helpers/src/lib.rs [hunk 5]
```diff
@@ -341,9 +341,12 @@ fn rewrite_url_encoded_map_keys(input: &str) -> (String, String) {
 fn rewrite_url_encoded_body(input: &str) -> String {
     let mut entries: Vec<(String, String)> = input
         .split("\n&")
+        .filter(|s| !s.is_empty())
```
Comment (rcoh) [inline rust-runtime/protocol-test-helpers/src/lib.rs:344]: not sure of the exact context, but you might want `s.trim().is_empty()`

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/AwsQuerySerializerGenerator.kt (hunk 2 above)
Comment (jdisanti) [reply codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/AwsQuerySerializerGenerator.kt:187]: Added comment.

File: rust-runtime/protocol-test-helpers/src/lib.rs (hunk 5 above)
Comment (jdisanti) [reply rust-runtime/protocol-test-helpers/src/lib.rs:344]: Fixed

File: rust-runtime/protocol-test-helpers/src/lib.rs (hunk 4 above)
Comment (jdisanti) [reply rust-runtime/protocol-test-helpers/src/lib.rs]: Definitely. Moved.

**Pull request merged**